Update core mass calculation and add tidal heating model of the soliton

yzhaoastro opened the pull request:
This PR introduces the implementation of core–halo mass relation time integration based on Chan et al. (2022):
Added support for evolving massCore using the time derivative formulation.
Verified consistency against the original expression (Eq. 15 in Chan et al. 2022): Maximum relative difference ~ 0.01% and Minimum difference ~0.0002%.

Soliton tidal heating model (Du et al. 2018):
This operator accumulates the tidal-heating source term from the FDM solitonic core.

These changes improve the fidelity of soliton core evolution and enable tidal-heating modeling for subhalos.

abensonca commented:
It looks like [`tests.dark_matter_profiles.fuzzy_dark_matter.exe`](https://github.com/galacticusorg/galacticus/actions/runs/17780308811/job/50544569723?pr=927) needs some updating to work with the updated code.

yzhaoastro commented on source/dark_matter_profiles_DMO.soliton_NFW.F90:
Yes, I’ll keep the random offset here, since I still need to check for a valid soliton radius solution within this file.

yzhaoastro commented on source/nodes.operators.physics.satellite.mass_loss.soliton.F90:
```diff
@@ -0,0 +1,229 @@
+    ! Set the rate to be integrated by Galacticus for the soliton core density evolution. From Du et al. (2018), equation (17) can be written as: d(log ρ_c) / dt = 2 Im(E)
+    ! Im(E) is given by their fitting formula (eq. 7) as a function of densityRatio = ρ_c / ρ̄_host(<r_orbit) and the orbital period T_orbit.
+    ! Here we evaluate the instantaneous rate 2*Im(E), which will be integrated over time to obtain Δlog ρ_c(t) = ∫ [2 Im(E)] dt.
+    call basic%floatRank0MetaPropertyRate(                             &
```
I’ve moved this out of the PR for now.

yzhaoastro commented on source/nodes.operators.physics.dark_matter_profile.soliton.F90:
```diff
@@ -0,0 +1,342 @@
+    class           (nodeComponentBasic                  ), pointer       :: basic
+
+    basic=>node%basic()
+    call basic%floatRank0MetaPropertyScale(self%densityCoreAccretionID, scaleRelative)
```
I’ve updated it to use 10% of the initialized massCore as the absolute scale.

abensonca commented:
> I also added `tests.dark_matter_profiles.Chan2022.F90`, which tests the core mass calculation by comparing the integrated result with the analytic formula from Chan et al. (2022).

I think this test is useful, but I don't think it's actually testing the integration. It initializes the soliton:
```
call nodeOperator_%nodeTreeInitialize(node_)
```
and then tests that the initialized core mass agrees with the analytic expectation. But, it doesn't evolve the soliton forward in time. 

Probably what you want to do is actually run the model that set up - i.e. do a full Galacticus run with it, like this:
```
./Galacticus.exe testSuite/parameters/darkMatterProfilesChan2022.xml
```
which will then initialize solitons in the halos defined in your `testSuite/parameters/fdmSingleTree.xml` file and then evolve them forward in time.

I think you'll then want to create a new `nodePropertyExtractor` class which can extract the properties that we want to test (e.g. core mass) so that they can be output. I think you'll want to add a parameter to the `darkMatterProfileDMOSolitonNFW` class to allow the amount of scatter to be set - this can default to the current value of `log10(1.5d0)`, but in your test you can set this to zero so that we can check if the correct mean behavior is reproduced. I think we should also make the halo mass in `fdmSingleTree.xml` grow so that we are testing that the integration correctly accounts for halo mass growth. 

Once that is done, you'll need a small Python code (use `testSuite/test-spheroid-power-law.py` as an example) which runs the model, reads the data from the output file, and tests the core mass values against the known correct answers. This will then verify that the integration works correctly.

yzhaoastro commented:
I added a new nodePropertyExtractor to output the relevant soliton properties (solitonMassCore, solitonZeta0, solitonZetaZ, solitonExpansionFactor, solitonMassParticle, solitonMassHalo) into the nodeData. I also introduced a configurable scatterLog parameter in the darkMatterProfileDMOSolitonNFW class, which defaults to 1.5 but can be set to 1.0 in the tests to disable scatter and check the mean relation.

The fdmSingleTree.xml file was updated so that the halo mass grows with time, ensuring that the integration is tested under evolving halo conditions. For validation, I created a Python script (test-soliton-coremass.py) that runs the model, reads the HDF5 outputs, and compares the integrated core mass against the analytic expectation from Chan et al. (2022). This provides an end-to-end check that the integration reproduces the expected core–halo relation.

abensonca reviewed (changes requested):
As we discussed - just one significant comment. I'm going to make some small changes to the new test code also, and will push an update for that once I'm finished.

yzhaoastro commented on source/nodes.operators.physics.dark_matter_profile.soliton.F90:
I’m using the global minimum approach. Following Hui et al. (2017), the minimum halo mass expected to form in FDM is Mh ~ 10^7*(m_b/10−22 eV)^(−3/2).
Then, using the result from Schive et al. (2014) that who show that the soliton core mass is about one quarter of the halo mass, i.e. M_c ~ M_h/4, I set the minimum core mass to roughly one quarter of this minimum halo mass.

abensonca commented:
I pushed a few more updates - two significant changes:

1. I changed the way the core mass scatter parameter is defined such that a value of 0.0 means no scatter (instead of a value of 1.0 meaning no scatter) as I think this is more intuitive.
2. Some of the properties that were being extracted by the `nodePropertyExtractorSoliton` class were already available in the output file (via other `nodePropertyExtractor`s, or in the `Parameters` group), so I reduced that class to provided just the core mass.
